docs/proposals/20191016-e2e-test-framework.md

```diff
@@ -0,0 +1,207 @@
+    - [As the release manager of a Cluster API provider, I want to be able to recommend new releases to users as soon as they come up](#as-the-release-manager-of-a-cluster-api-provider-i-want-to-be-able-to-recommend-new-releases-to-users-as-soon-as-they-come-up)
+    - [As a user of Cluster API, I want to be sure new releases I’m installing won’t cause regressions.](#as-a-user-of-cluster-api-i-want-to-be-sure-new-releases-im-installing-wont-cause-regressions)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Behaviors to test](#behaviors-to-test)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+- [Upgrade Strategy](#upgrade-strategy)
+- [Additional Details](#additional-details)
+  - [Test Plan](#test-plan)
+  - [Version Skew Strategy [optional]](#version-skew-strategy-optional)
+- [Implementation History](#implementation-history)
+
+<!-- END doctoc generated TOC please keep comment here to allow auto update -->
+
+## Glossary
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+If this proposal adds new terms, or defines some, make the changes to the book's glossary when in PR stage.
+
+Basic e2es:
+The Reconcilers are exercised by creating a cluster and deleting a cluster
+Comprehensive e2es:
+Many of the components and behaviors of Cluster API are exercised.
+Creates a number of clusters and tears them down.
+
+## Summary
+
+Cluster API's providers could benefit from having a set of generic behavioral e2e tests. Most providers have some form
+of end-to-end testing but they are not uniform and do not all test the same behavior that Cluster API promises.
+
+A pluggable set of behavioral tests for Cluster API Providers would prove beneficial for the project health as a whole.
+
+As stated in the non-goals, this proposal does not intend to define the conformance requirements for Cluster API providers.
+
+## Motivation
+
+Every infrastructure and bootstrap provider wants to know if it "works", but testing today is a manual process.
+Developers test one or two well-trod scenarios if testing at all before merging PRs. This test suite, when properly
+configured, would provide a baseline signal for any given provider for more scenarios than any single developer would
+reasonably spend time testing.
+
+A bonus to creating a pluggable suite of behaviors is the consolidation of expectations in one place. The test suite is
+a direct reflection of the core provider and thus is versioned along with the cluster-api repository.
+
+### Goals
+
+* Define sets of tests that all providers can pass
+  * End-to-end tests
+  * (optional) Unit tests
+* Build a test suite as a go library that providers can import, supply their own types/objects that satisfy whatever interface is required of the tests
```

Comment:
Will these tests be run by the providers or will it be run in the CAPI repo?
I think there are benefits doing both:
- As a CAPI dev, I want to ensure that my changes don't break assumptions made by a subset of providers.
- As a provider dev, I would also want to ensure that my assumptions align w/ the premises agreed across the community.

Reply:
Answered here:
https://github.com/kubernetes-sigs/cluster-api/pull/1571/files#r337686384
> Yes, it does. Extrapolating this to answer my other question, as a provider dev, I will be running these tests and reporting back CAPI changes that break my provider implementation.